fearless/ApplicationLayer/Services/WalletBalanceSubscriptionAdapter.swift

```diff
@@ -0,0 +1,470 @@
+    }
+
+    private func fetchAllMetaAccounts() {
+        let metaAccountsOperation = metaAccountRepository.fetchAllOperation(with: RepositoryFetchOptions())
+        let chainsOperation = fetchChainsOperation()
+
+        metaAccountsOperation.addDependency(chainsOperation)
+
+        metaAccountsOperation.completionBlock = { [weak self] in
+            guard let metaAccountsResult = metaAccountsOperation.result else {
+                self?.deliverOnCompletion(.failure(WalletBalanceError.accountMissing))
+                return
+            }
+            guard let chainsResult = chainsOperation.result else {
+                self?.deliverOnCompletion(.failure(WalletBalanceError.chainsMissing))
+                return
+            }
+
+            switch (metaAccountsResult, chainsResult) {
+            case let (.success(wallets), .success(chains)):
+                self?.handle(wallets, chains)
+            case let (.failure(error), _):
+                self?.deliverOnCompletion(.failure(error))
+            case let (_, .failure(error)):
+                self?.deliverOnCompletion(.failure(error))
+            }
+        }
+
+        operationQueue.addOperations([chainsOperation, metaAccountsOperation], waitUntilFinished: false)
+    }
+
+    private func subscribeToAccountInfo(
+        for wallets: [MetaAccountModel],
+        _ chainAssets: [ChainAsset]
+    ) {
+        wallets.forEach { wallet in
+            let accountInfoSubscriptionAdapter = AccountInfoSubscriptionAdapter(
+                walletLocalSubscriptionFactory: WalletLocalSubscriptionFactory.shared,
+                selectedMetaAccount: wallet
+            )
+
+            accountInfosAdapters[wallet.identifier] = accountInfoSubscriptionAdapter
+            accountInfoSubscriptionAdapter.subscribe(chainsAssets: chainAssets, handler: self)
+        }
+    }
+
+    private func subscribeToPrices(for chains: [ChainModel]) {
+        let pricesIds = chains.compactMap { $0.assets.compactMap(\.asset.priceId) }.reduce([], +)
+        pricesProvider = subscribeToPrices(for: pricesIds)
+    }
+
+    private func fetchChainsOperation() -> BaseOperation<[ChainModel]> {
+        chainRepository.fetchAllOperation(with: RepositoryFetchOptions())
+    }
+
+    private func deliverOnCompletion(_ result: WalletBalancesResult) {
+        dispatchInQueueWhenPossible(deliverQueue) { [weak self] in
+            self?.delegate?.handle(result: result)
+        }
+    }
+}
+
+// MARK: - EventVisitorProtocol
+
+extension WalletBalanceSubscriptionAdapter: EventVisitorProtocol {
+    func processMetaAccountChanged(event: MetaAccountModelChangedEvent) {
+        if let index = metaAccounts.firstIndex(where: { $0.metaId == event.account.metaId }) {
+            metaAccounts[index] = event.account
+        }
+        buildBalance()
+    }
+
+    func processChainsUpdated(event: ChainsUpdatedEvent) {
+        let updatedChainAssets = event.updatedChains.map(\.chainAssets).reduce([], +)
```

Comment:
you can make this code more fast, beautiful, and stable in case you will store chainAssets as [ChainAssetId: ChainAsset] hashmap instead of [ChainAsset] array. accountInfos already stored as hashmap